UnitTests/UnitTests.UWP/UI/Controls/Test_ImageEx.cs

```diff
@@ -0,0 +1,91 @@
+
+        [TestMethod]
+        public async Task SetSourceToOpenedBitmapImage()
+        {
+            await App.DispatcherQueue.EnqueueAsync(async () =>
+            {
+                var bitmapImage = new BitmapImage();
+
+                using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(@"UnitTests.UWP.Assets.StoreLogo.embeded.png");
+                using var memStream = new MemoryStream();
+                await stream.CopyToAsync(memStream);
+                memStream.Seek(0, SeekOrigin.Begin);
+                await bitmapImage.SetSourceAsync(memStream.AsRandomAccessStream());
+
+                var imageLoader = new ImageEx();
+
+                await SetTestContentAsync(imageLoader);
+
+                Assert.AreEqual("Unloaded", GetCurrentState(imageLoader));
+
+                var imageOpendedCallCount = 0;
+
+                imageLoader.ImageExOpened += (s, e) =>
+                {
+                    imageOpendedCallCount++;
+                    Assert.AreEqual("Loaded", GetCurrentState(imageLoader));
+                };
+
+                imageLoader.Source = bitmapImage;
+
+                Assert.AreEqual(1, imageOpendedCallCount, "{0} should only be called once", nameof(ImageEx.ImageExOpened));
+            });
+        }
+
+        [TestMethod]
+        [DataRow(ImageString)]
+        [DataRow(@"ms-appx:///Assets/StoreLogo.png")]
+        public async Task SetSourceToUri(string uri)
+        {
+            await App.DispatcherQueue.EnqueueAsync(async () =>
+            {
+                var imageLoader = new ImageEx();
+
+                await SetTestContentAsync(imageLoader); 
+
+                Assert.AreEqual("Unloaded", GetCurrentState(imageLoader));
+
+                var imageOpendedCallCount = 0;
+                imageLoader.ImageExOpened += (s, e) =>
+                {
+                    imageOpendedCallCount++;
+                    Assert.AreEqual("Loaded", GetCurrentState(imageLoader));
+                };
+
+                imageLoader.Source = new Uri(uri);
+
+                // TODO (2021.05.11): Test in a more deterministic way.
+                // Setting source causes some async code to trigger and
+                // we have no way to await or handle its complementation regardless of the result.
```

Comment:
Couldn't we do that by using a [`TaskCompletionSource<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1) and setting its result from the handler for `ImageExOpened`, so that we'd always get past this point once the image has actually loaded?
We might also consider using a timeout for the wait in that case if we also wanted to avoid blocking the test indefinitely if the image fails to open, but was wondering if at least this approach would solve this determinism issue you mentioned here? 🤔

Reply:
Yeah, it's also weird as we care about how many times it's called as we only want to see it called once on success. I believe the previous version would call it twice in some of these scenarios, which is why I had removed the extra calls to the event, but didn't realize that would break if an already loaded image was passed in.

We were originally hoping to use the FluentAssertion library helpers here, but [that's not supported for UWP](https://github.com/fluentassertions/fluentassertions/issues/1557).